### PR TITLE
vtk: Bump to 8.2.0

### PR DESCRIPTION
**Pull request by rubendibattista**
###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.x
Xcode 8.x

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by rubendibattista** [comment]
PS: I have a proposal

Why don't we keep `vtk` and `paraview` in the same Portfile (`vtk` is a subport), where `vtk` is checked out to the right commit associated to the `paraview` version.

With the approriate handling of `cmake` arguments, we might use the Paraview's `PARAVIEW_USE_EXTERNAL_VTK` to build VTK just once...

**Comment by mf2k** [comment]
This PR appears to be missing the patch file pa`tch-libproj.diff`. 


```
--->  Attempting to fetch patch-libproj.diff from http://www.vtk.org/files/release/8.2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   236  100   236    0     0   4379      0 --:--:-- --:--:-- --:--:--  4452
 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
DEBUG: Fetching distfile failed: The requested URL returned error: 404 Not Found

```

**Comment by ra1nb0w** [comment]
Fails to build see [main.log](https://github.com/macports/macports-ports/files/4466270/main.log)

> PS: I have a proposal

Seems a good idea seen that they are related (but I used paraview many many years ago). Open a new PR with your proposal.

**Comment by rubendibattista** [comment]
I actually don't understand what's the problem with `libogg`. I tried to let it compile its own vendored version, but it still fails.

On my machine it compiled with libogg provided by Macports.

**Comment by ra1nb0w** [comment]
you should use the macports libogg. why don't?
error: when linking `IO/Movie` the linker didn't find the ogg symbols; normaly because it doesn't find the library or it is not included in the `target_link_libraries`

**Comment by rubendibattista** [comment]
Ehi @ra1nb0w, actually `libogg` is a vendored dependency of VTK that isn't supposed to be "externalized".  As for `libProj`. I guess because of some modifications VTK's team did without upstreaming them? You can understand this from the `CMakeLists.txt` associated to the dependency. 

For example, let's take [HDF5](https://github.com/Kitware/VTK/blob/master/ThirdParty/hdf5/CMakeLists.txt), in particular [these lines](https://github.com/Kitware/VTK/blob/master/ThirdParty/hdf5/CMakeLists.txt#L18-L30). They suggest that HDF5 can be provided externally. 

Meanwhile, if we check `ogg` or `libproj`, we get a [this](https://github.com/Kitware/VTK/blob/master/ThirdParty/ogg/CMakeLists.txt). That's an indication they it's assumed to be built from the vendored version. 

Now, the problem is, even if I let VTK build `ogg` from the vendored version, it does not build.

**Comment by ra1nb0w** [comment]
[this](https://github.com/ra1nb0w/macports-ports/commit/3c6cf4e8eeeca6f571192ff9caa0412defdb0541) build fine with system ogg. try it out.
What is the problem with `SYSTEM_DIY2`? Why do you need a patch to find `pugixml`?

note: pay attention on spaces (read the first line)

**Comment by rubendibattista** [comment]
@ra1nb0w Ehi, thanks for the patch of the Movie IO. 

The patch for `pugixml` is needed to let VTK find it, since it is another of those dependencies assumed to be vendor-only. 

For the DIY2, it was not building. I don't remember the error tho. I also tried to use the `port:diy` instead, without luck.

**Comment by ra1nb0w** [comment]
> The patch for `pugixml` is needed to let VTK find it, since it is another of those dependencies assumed to be vendor-only.

strange that it is not already in. is there a cmake parameter? probably you should open an issue in VTK

> For the DIY2, it was not building. I don't remember the error tho. I also tried to use the `port:diy` instead, without luck.

you should use `port:diy2`. I try to build it with `+mpich` and see the error.

> PS: Which spaces?

the ones in the new python section.

**Comment by ra1nb0w** [comment]
ps. try out my patch if it works fine with the others software.

**Comment by rubendibattista** [comment]
Yup. I'm currently building with your IO patch on my machine (and I also restored DIY2). Let's see...

> strange that it is not already in. is there a cmake parameter? probably you should open an issue in VTK

I don't think there's since, I think, is managed internally and it does not expose the parametrs for ThirdParty external libraries as usual...

**Comment by ra1nb0w** [comment]
is it ok with my patch?
please copy and past my portfile and files/ in this PR otherwise I will commit my version and I cite your name as contributor.

**Comment by rubendibattista** [comment]
It looks like there's a problem with the Xcode 7.3 build. Should we try to fix it on there?

**Comment by rubendibattista** [comment]
Yes, with your patch it's working :)

**Comment by ra1nb0w** [comment]
first copy my portfile and files/. Your is different!
then you can investigate why of this error (from xcode 7.2 log file)
```
In file included from /opt/local/var/macports/build/_Users_travis_build_macports_macports-ports_graphics_vtk/vtk/work/build/ThirdParty/jsoncpp/vtk_jsoncpp.h:25:
In file included from /opt/local/include/json/json.h:11:
In file included from /opt/local/include/json/reader.h:11:
/opt/local/include/json/value.h:680:32: error: inline declaration of 'as<bool>' follows non-inline definition
template <> inline bool Value::as<bool>() const { return asBool(); }
```
Probably [compiler](https://trac.macports.org/wiki/UsingTheRightCompiler) or you need to force c++11 at cmake arguments

**Comment by rubendibattista** [comment]
It looks like a [bug of the compiler](https://stackoverflow.com/questions/45284548/deleted-template-function-works-on-gcc-but-not-on-clang)

**Comment by rubendibattista** [comment]
> first copy my portfile and files/. Your is different!
> then you can investigate why of this error (from xcode 7.2 log file)
> 
> ```
> In file included from /opt/local/var/macports/build/_Users_travis_build_macports_macports-ports_graphics_vtk/vtk/work/build/ThirdParty/jsoncpp/vtk_jsoncpp.h:25:
> In file included from /opt/local/include/json/json.h:11:
> In file included from /opt/local/include/json/reader.h:11:
> /opt/local/include/json/value.h:680:32: error: inline declaration of 'as<bool>' follows non-inline definition
> template <> inline bool Value::as<bool>() const { return asBool(); }
> ```
> 
> Probably [compiler](https://trac.macports.org/wiki/UsingTheRightCompiler) or you need to force c++11 at cmake arguments

Ehi I know they're different, but: 

```
                    -DVTK_USE_SYSTEM_OGG:BOOL=ON \
                    -DVTK_USE_SYSTEM_THEORA:BOOL=ON \
```

these are not needed. `-DVTK_USE_SYSTEM_LIBRARIES` enables for them all. I re-established to use also the external `ogg` library provided by the system, as in your Portfile (that I forgot to apply to this PR).

**Comment by ra1nb0w** [comment]
please, add a comment before `compiler.blacklist-append` otherwise we can't understand why you are adding this. Anyway, this is a bug because the documentation states that the minimum macOS supported is 10.7.

**Comment by rubendibattista** [comment]
> please, add a comment before `compiler.blacklist-append` otherwise we can't understand why you are adding this. Anyway, this is a bug because the documentation states that the minimum macOS supported is 10.7.

The bug it's a bug of the compiler and  comes from `jsoncpp` installed by Macports, not from VTK: Macports ships `v1.9.2` while `VTK` vendors `v1.8.4`. I think that blacklisting should be inserted into `jsoncpp` `Portfile` instead of `VTK`, what do you think @ra1nb0w?

**Comment by rubendibattista** [comment]
How does it work with the compilers? If one compiler is blacklisted for a dependency, it is also blacklisted for the dependent?

**Comment by rubendibattista** [comment]
I pushed a stricter blacklisting on `jsoncpp`. Let's see.

**Comment by ra1nb0w** [comment]
if it works: jsoncpp commit must not touch vtk portfile. squash all vtk changes in one commit that should be after the jsoncpp commit.

**Comment by rubendibattista** [comment]
It looks like the blacklisting doesn't get propagated to dependents. I think it should. Nonetheless I applied the same blacklisting to `vtk`.

**Comment by ra1nb0w** [comment]
the blacklist is per port.
@michaelld can you check the jsoncpp commit. thank you.

**Comment by michaelld** [comment]
> the blacklist is per port.
> @michaelld can you check the jsoncpp commit. thank you.

Working on it. Pretty sure it requires AppleClang >= 900, so blocking < 800 will fail for OSX that has AppleClang 900 installed. Will test shortly ...

File: devel/jsoncpp/Portfile [hunk 1]
```diff
@@ -9,7 +9,7 @@ PortGroup           compiler_blacklist_versions 1.0
 compiler.cxx_standard 2011
 
 # some older Clang say they support C++11 when they don't
-compiler.blacklist-append {clang < 900}
+compiler.blacklist-append {clang < 800} {macports-clang-3.[0-8]}
```
Comment (michaelld) [inline devel/jsoncpp/Portfile:12]: this needs to be `< 900` ... I just tested ...
```
$ clang -v
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
and here's the logfile for your verification ...
[jsoncpp_clang_800.log.txt](https://github.com/macports/macports-ports/files/4476897/jsoncpp_clang_800.log.txt)

if I'm missing something, please tell me what to check or test or change & I'll try it out.

NOTE: This is an actual MacBook running OSX 10.11 ...

Comment (rubendibattista) [reply devel/jsoncpp/Portfile:12]: Ehi thanks @michaellass, so the blacklist of the `macports-clang` up to 3.8 version is not needed?

Comment (kencu) [reply devel/jsoncpp/Portfile]: that fits .... clang 800 is approximately macports-clang-3.8, clang 900 = macports-clang-3.9.

so blacklisting clang < 900 means about the same as blacklisting macports-clang-3.[0-8].

**Comment by ra1nb0w** [comment]
Move ```PortGroup           cmake 1.0``` to ```PortGroup           cmake 1.1``` and remove the fourth empty line.

sorry for the (probably) stupid question: but why can't we use
```compiler.cxx_standard 2011```
and avoid
```
PortGroup           cxx11 1.1
PortGroup           compiler_blacklist_versions 1.0
```
?

**Comment by rubendibattista** [comment]
> Move `PortGroup cmake 1.0` to `PortGroup cmake 1.1` and remove the fourth empty line.
> 
> sorry for the (probably) stupid question: but why can't we use
> `compiler.cxx_standard 2011`
> and avoid
> 
> ```
> PortGroup           cxx11 1.1
> PortGroup           compiler_blacklist_versions 1.0
> ```
> 
> ?

The blacklist is still needed since those compilers, even if they advertize to support C++11, they do not completely.

I will update the rest you ask, that seems reasonable

**Comment by rubendibattista** [comment]
I think it's good now. Also on Xcode 7.3 it timed-out the Travis run (can't we switch everything to Azure?)

**Comment by ra1nb0w** [comment]
LGTM. Try to merge. thank you.
if you would like to add yourself as maintainer open a new PR with "vtk: add myself as maintainer"

**Pull request merged**